nixos/modules/services/audio/librespot.nix

```diff
@@ -0,0 +1,220 @@
+{ config, pkgs, lib, ... }:
+
+with builtins;
+with lib;
+with types;
+
+let
+  cfg = config.services.librespot;
+in
+{
+  options = {
+    services.librespot = {
+      enable = mkEnableOption "librespot: Open Source Spotify client library";
+
+      cache = mkOption {
+        type = nullOr str;
+        default = null;
+        description = "Path to a directory where files will be cached.";
+      };
+
+      enableAudioCache = mkOption {
+        type = bool;
+        default = true;
+        description = "Enable caching of the audio data.";
+      };
+
+      name = mkOption {
+        type = str;
+        description = "Device name";
+      };
+
+      deviceType = mkOption {
+        type = enum [ "unknown" "computer" "tablet" "smarthphone" "speaker" "tv" "avr" "stb" "audiodongle" ];
+        default = "speaker";
+        description = "The displayed device type in Spotify clients. Can be unknown, computer, tablet, smartphone, speaker, tv, avr (Audio/Video Receiver), stb (Set-Top Box), and audiodongle. Defaults to speaker";
+      };
+
+      bitrate = mkOption {
+        type = nullOr int;
+        default = null;
+        description = "Bitrate (96, 160 or 320). Defaults to 160";
+      };
+
+      onevent = mkOption {
+        type = nullOr path;
+        default = null;
+        description = "The path to a script that gets run when one of librespot's events is triggered.";
+      };
+
+      verbose = mkOption {
+        type = bool;
+        default = false;
+        description = "Enable verbose output";
+      };
+
+      username = mkOption {
+        type = nullOr str;
+        default = null;
+        description = "Username to sign in with";
+      };
+
+      password = mkOption {
+        type = nullOr str;
+        default = null;
+        description = "Password to sign in with";
+      };
```

Comment:
secrets should not be specified as a string, because they would end up in the nix-store. better to have them be read from a path.